File: bindings/cpp/test/test_agg.cpp
#include "hocdb_cpp.h"
#include <iostream>
#include <cassert>
#include <vector>
#include <filesystem>

struct TradeData {
    int64_t timestamp;
    double value;
};

int main() {
    try {
        std::string test_dir = "b_cpp_test_data/agg";
        if (std::filesystem::exists(test_dir)) {
            std::filesystem::remove_all(test_dir);
        }
        std::filesystem::create_directories(test_dir);

        std::vector<hocdb::Field> schema = {
            {"timestamp", HOCDB_TYPE_I64},
            {"value", HOCDB_TYPE_F64}
        };

        hocdb::Database db("TEST_CPP_AGG", test_dir, schema);

        std::cout << "Appending data..." << std::endl;
        db.append(TradeData{100, 10.0});
        db.append(TradeData{200, 20.0});
        db.append(TradeData{300, 30.0});
        db.flush();

        std::cout << "Testing getLatest..." << std::endl;
        auto [val, ts] = db.getLatest(1); // value index = 1
        std::cout << "Latest: value=" << val << ", timestamp=" << ts << std::endl;
        
        assert(val == 30.0);
        assert(ts == 300);

        std::cout << "Testing getStats..." << std::endl;
        auto stats = db.getStats(0, 400, 1);
        std::cout << "Stats: min=" << stats.min << ", max=" << stats.max << ", sum=" << stats.sum 
                  << ", count=" << stats.count << ", mean=" << stats.mean << std::endl;

        assert(stats.count == 3);
        assert(stats.min == 10.0);
        assert(stats.max == 30.0);
        assert(stats.sum == 60.0);
        assert(stats.mean == 20.0);

        std::cout << "C++ Aggregation Test Passed!" << std::endl;
        return 0;
    } catch (const std::exception& e) {
        std::cerr << "Test failed: " << e.what() << std::endl;
        return 1;
    }
}

File: bindings/bun/test_agg.ts
import { HOCDB } from "./index.ts";
import { rmSync, existsSync } from "fs";

import { join } from "path";

const TEST_DIR = join(import.meta.dir, "..", "..", "b_bun_test_data");
const TICKER = "TEST_BUN_AGG";

if (existsSync(TEST_DIR)) {
    rmSync(TEST_DIR, { recursive: true, force: true });
}

const schema = [
    { name: "timestamp", type: "i64" as const },
    { name: "value", type: "f64" as const }
];

const db = new HOCDB(TICKER, TEST_DIR, schema, {
    max_file_size: 1024 * 1024,
    flush_on_write: true
});

console.log("Appending data...");
db.append({ timestamp: 100n, value: 10.0 });
db.append({ timestamp: 200n, value: 20.0 });
db.append({ timestamp: 300n, value: 30.0 });

console.log("Testing getLatest...");
const latest = db.getLatest(1); // value index = 1
console.log("Latest:", latest);

if (latest.value !== 30.0 || latest.timestamp !== 300n) {
    throw new Error(`getLatest failed: expected {value: 30.0, timestamp: 300}, got ${JSON.stringify(latest, (key, value) => typeof value === 'bigint' ? value.toString() : value)}`);
}

console.log("Testing getStats...");
const stats = db.getStats(0n, 400n, 1);
console.log("Stats:", stats);

if (stats.count !== 3n) throw new Error(`Count mismatch: expected 3, got ${stats.count}`);
if (stats.min !== 10.0) throw new Error(`Min mismatch: expected 10.0, got ${stats.min}`);
if (stats.max !== 30.0) throw new Error(`Max mismatch: expected 30.0, got ${stats.max}`);
if (stats.sum !== 60.0) throw new Error(`Sum mismatch: expected 60.0, got ${stats.sum}`);
if (stats.mean !== 20.0) throw new Error(`Mean mismatch: expected 20.0, got ${stats.mean}`);

db.close();
console.log("Bun Aggregation Test Passed!");
